### ISSUE TITLE
Adopt Semver

### ISSUE THREAD

**Issue by silverwind** [go-gitea/gitea#27574]
### Feature Description

I think it would be good to adopt [semver](https://semver.org/) for gitea versioning, e.g. the current minor releases would become major ones, signifying the breaking changes contained in these releases. So next version would be 2.0.0, a feature after that would be 2.1.0 and a bugfix-only release after that 2.1.1.

**Comment by denyskon** [comment]
If I understand correctly, releasing major versions in Go means effectively creating a new module every time... this does sound like something that would make our workflow harder

**Comment by silverwind** [comment]
Yes, some automation would be needed to rewrite the go files, but imho just because the language makes it hard to do major bumps, it's no excuse to not do it.

**Comment by techknowlogick** [comment]
pinging @jolheiser as he had some thoughts on this as well.

**Comment by lunny** [comment]
1.21.x -> 21.x.x
1.22.x -> 22.x.x

**Comment by wxiaoguang** [comment]
I ever wrote some drafts (when Gitea is at 1.19 stage), maybe it's time to share them 😁 

----

At the moment, although Gitea's version is "1.19" or "1.20", it's not a semantic version.

There are some problems:

1. It confuses with "Golang" version (that's also 1.19/1.20 ....)
2. It make people suppose that "1.x" is the same, however, each "minor" release changes a lot
3. It's difficult to make people know when a version is released.


# Possible alternatives

## Date versioning system

That's also what many large applications do:

For example: `Year.ReleaseNumber.PatchFix-suffix`

* 2023.1.0-rc0
* 2023.1.1
* 2023.4.5

* It could still be parsed by semantic version.
* The same `Year.ReleaseNumber`  doesn't break. 
* In year 2024, the release "2023.4" could still get new patch-fix releases


## Switch to semantic version

Just like Java / Chrome, increase the major number: `1.19.2` -> `19.2`

**Comment by jolheiser** [comment]
There is no reason we would need to change the Go imports, as I detailed in the Discord when I last proposed this same thing at the beginning of this year.

Gitea is not installable via `go install` due to our build process, nor is it a valid dependency such that anyone would reasonably `go get` it, so there is no reason (aside from Go convention) to version the module in compliance with SIV.

**Comment by silverwind** [comment]
One benefit semver brings is being able to picture feature-releases in the version which previously was not possible with just two numbers. Not sure if we'd really need it thought, but I guess it's good to have an option to make a minor release if a feature warrants it.

**Comment by denyskon** [comment]
> There is no reason we would need to change the Go imports, to the best of my knowledge.
> 
> Gitea is not installable via `go install` due to our build process, nor is it a valid dependency such that anyone would reasonably `go get` it, so there is no reason (aside from Go convention) to version the module in compliance with SIV.
> 
> EDIT: Due to the `replace` directives in our module, neither `go get` or `go install` will work (correctly) regardless of the above reasons.

Well we discovered that gitness imports some of our internal packages directly iirc. But is it something we should be bothered about?

**Comment by jolheiser** [comment]
> But is it something we should be bothered about?

No, and even then they can still import it, it will just be marked `incompatible` in their module.

**Comment by silverwind** [comment]
One problem is golang makes it far too hard to have proper semver. Ever wondered why almost all go projects are at v1 while having done many breaking changes? golang's stupid requirement of having the version in the module path is the reason:

https://donatstudios.com/Go-v2-Modules
https://github.com/golang/go/issues/31543

**Comment by jolheiser** [comment]
> One problem is golang makes it far too hard to have proper semver. Ever wondered why almost all go projects are at v1 while having done many breaking changes? golang's stupid requirement of having the version in the module path is the reason:
> 
> [donatstudios.com/Go-v2-Modules](https://donatstudios.com/Go-v2-Modules) [golang/go#31543](https://github.com/golang/go/issues/31543)

This isn't a problem for us, see https://github.com/go-gitea/gitea/issues/27574#issuecomment-1757729006

**Comment by silverwind** [comment]
> > One problem is golang makes it far too hard to have proper semver. Ever wondered why almost all go projects are at v1 while having done many breaking changes? golang's stupid requirement of having the version in the module path is the reason:
> > [donatstudios.com/Go-v2-Modules](https://donatstudios.com/Go-v2-Modules) [golang/go#31543](https://github.com/golang/go/issues/31543)
> 
> This isn't a problem for us, see [#27574 (comment)](https://github.com/go-gitea/gitea/issues/27574#issuecomment-1757729006)

Would we still need to rewrite all internal imports on every major bump or could we circumvent that with a `replace` directive in go.mod?

**Comment by jolheiser** [comment]
Neither would be required. The module version only matters for other Go ecosystem things to interact with us, which is already impossible due to other factors or is extremely unlikely.

**Comment by silverwind** [comment]
> Neither would be required. The module version only matters for other Go ecosystem things to interact with us, which is already impossible due to other factors or is extremely unlikely.

I thought golang would basically break as soon as module version does not match git version, but I guess go itself can and must work without a git dependency, so therefor it should never read the git tag info.

> That is to say, yes we would technically not be following Go's versioning for SIV,

What is SIV? Sorry, not familiar with that term.

**Comment by jolheiser** [comment]
SIV is semantic import versioning, e.g. requiring the module path to include `/vN`

Go works with many VCS, not just git, so it can never truly rely on git-specific information. We also don't use any of the internal VCD info, so no breakage there.

The things that break are essentially non-issues for us because we aren't a library and because we already aren't feasible for other reasons.

At this point Gitea is only part of the Go ecosystem in language, more or less. Nothing can feasibly depend on our module, and we require a build step so install and run are out.

**Comment by jolheiser** [comment]
We could (and should) test it out beforehand in a duplicate repo to iron out release kinks, etc.
